File: apps/android/app/src/main/java/com/sennnen/mav/MavSnapshot.kt
package com.sennnen.mav

import org.json.JSONObject

data class MavSnapshot(
    val coreVersion: String,
    val storageSchema: Int,
    val revision: ULong,
    val asOfUnixMs: Long,
    val connectionState: String,
    val deviceName: String?,
    val batteryPercent: Int?,
    val charging: Boolean?,
    val lastSampleUnixMs: Long?,
    val currentBpm: Int?,
    val meanMilliBpm: Int?,
    val inRangeSamples: Int?,
    val excludedSamples: Int?,
    val prv: MavPrv?,
    val prvUnavailableReason: String?,
    val recoveryUnavailableReason: String?,
    val hash: String,
)

/**
 * The admitted time-domain variability read model. WHOOP intervals are PPG-derived, so the core
 * labels the result `pulse_rate_variability`; it is presented as PRV, never as ECG HRV
 * (docs/analytics.md).
 */
data class MavPrv(
    val label: String,
    val intervalSource: String,
    val meanIntervalMicros: Long,
    val rmssdMicros: Long,
    val sdnnMicros: Long,
    val nn50Count: Int,
    val pnn50MilliPercent: Long,
    val intervalCount: Int,
    val excludedIntervalCount: Int,
    val algorithm: String,
    val algorithmVersion: String,
    val provenanceId: Long,
)

object MavSnapshotDecoder {
    fun decode(json: String, hash: String): MavSnapshot {
        val root = JSONObject(json)
        val schema = root.getString("schema")
        require(schema == "host-snapshot/v1") { "unsupported snapshot schema: $schema" }
        val connection = root.getJSONObject("connection")
        val session = root.optJSONObject("session")
        val analytics = root.optJSONObject("analytics")
        return MavSnapshot(
            coreVersion = root.getString("core_version"),
            storageSchema = root.getInt("storage_schema"),
            revision = root.getLong("revision").toULong(),
            asOfUnixMs = root.getLong("as_of_unix_ms"),
            connectionState = connection.getString("state"),
            deviceName = connection.optNullableString("display_name"),
            batteryPercent = connection.optNullableInt("battery_percent"),
            charging = connection.optNullableBoolean("charging"),
            lastSampleUnixMs = connection.optNullableLong("last_sample_unix_ms"),
            currentBpm = session?.optNullableInt("current_bpm"),
            meanMilliBpm = session?.optNullableInt("mean_milli_bpm"),
            inRangeSamples = session?.getInt("in_range_samples"),
            excludedSamples = session?.getInt("excluded_samples"),
            prv = analytics?.let(::decodePrv),
            prvUnavailableReason = analytics?.unavailableReason("time_domain_hrv", "PRV"),
            recoveryUnavailableReason = analytics?.unavailableReason("recovery", "Recovery"),
            hash = hash,
        )
    }

    private fun decodePrv(analytics: JSONObject): MavPrv? {
        if (analytics.isNull("variability_label")) return null
        return MavPrv(
            label = analytics.getString("variability_label"),
            intervalSource = analytics.getString("interval_source"),
            meanIntervalMicros = analytics.getLong("mean_interval_micros"),
            rmssdMicros = analytics.getLong("rmssd_micros"),
            sdnnMicros = analytics.getLong("sdnn_micros"),
            nn50Count = analytics.getInt("nn50_count"),
            pnn50MilliPercent = analytics.getLong("pnn50_milli_percent"),
            intervalCount = analytics.getInt("interval_count"),
            excludedIntervalCount = analytics.getInt("excluded_interval_count"),
            algorithm = analytics.getString("algorithm"),
            algorithmVersion = analytics.getString("algorithm_version"),
            provenanceId = analytics.getLong("provenance_id"),
        )
    }
}

private fun JSONObject.unavailableReason(analytic: String, displayName: String): String? {
    val availability = optJSONArray("availability") ?: return null
    for (index in 0 until availability.length()) {
        val item = availability.getJSONObject(index)
        if (item.getString("analytic") != analytic || item.getBoolean("available")) continue
        val reason = item.optJSONObject("reason") ?: return "Unavailable"
        return when (reason.getString("kind")) {
            "algorithm_not_admitted" -> "$displayName model not admitted"
            "missing_streams" -> {
                val streams = reason.getJSONArray("streams")
                val names = (0 until streams.length()).joinToString { streams.getString(it) }
                "Needs $names"
            }
            else -> "Unavailable"
        }
    }
    return null
}

private fun JSONObject.optNullableString(name: String): String? =
    if (isNull(name)) null else getString(name)

private fun JSONObject.optNullableInt(name: String): Int? =
    if (isNull(name)) null else getInt(name)

private fun JSONObject.optNullableLong(name: String): Long? =
    if (isNull(name)) null else getLong(name)

private fun JSONObject.optNullableBoolean(name: String): Boolean? =
    if (isNull(name)) null else getBoolean(name)

File: apps/ios/Maverick/MavSnapshot.swift
import Foundation

struct MavSnapshot: Equatable, Sendable {
  let coreVersion: String
  let storageSchema: Int
  let revision: UInt64
  let asOfUnixMs: Int64
  let connectionState: String
  let deviceName: String?
  let batteryPercent: Int?
  let charging: Bool?
  let onWrist: Bool?
  let lastSampleUnixMs: Int64?
  let currentBpm: Int?
  let meanMilliBpm: Int?
  let inRangeSamples: Int?
  let excludedSamples: Int?
  let prv: MavPrv?
  let prvUnavailableReason: String?
  let recoveryUnavailableReason: String?
  let hash: String
}

/// The admitted time-domain variability read model. WHOOP intervals are PPG-derived, so the core
/// labels the result `pulse_rate_variability`; it is presented as PRV, never as ECG HRV
/// (docs/analytics.md).
struct MavPrv: Equatable, Sendable {
  let label: String
  let intervalSource: String
  let meanIntervalMicros: Int64
  let rmssdMicros: Int64
  let sdnnMicros: Int64
  let nn50Count: Int
  let pnn50MilliPercent: Int64
  let intervalCount: Int
  let excludedIntervalCount: Int
  let algorithm: String
  let algorithmVersion: String
  let provenanceId: Int64
}

enum MavSnapshotDecoder {
  static func decode(json: String, hash: String) throws -> MavSnapshot {
    let root = try JSONDecoder().decode(HostSnapshot.self, from: Data(json.utf8))
    guard root.schema == "host-snapshot/v1" else { throw MavPresentationError.unsupportedSchema }
    return MavSnapshot(
      coreVersion: root.coreVersion,
      storageSchema: root.storageSchema,
      revision: root.revision,
      asOfUnixMs: root.asOfUnixMs,
      connectionState: root.connection.state,
      deviceName: root.connection.displayName,
      batteryPercent: root.connection.batteryPercent,
      charging: root.connection.charging,
      onWrist: root.connection.onWrist,
      lastSampleUnixMs: root.connection.lastSampleUnixMs,
      currentBpm: root.session?.currentBpm,
      meanMilliBpm: root.session?.meanMilliBpm,
      inRangeSamples: root.session?.inRangeSamples,
      excludedSamples: root.session?.excludedSamples,
      prv: try root.analytics.flatMap(prv(from:)),
      prvUnavailableReason: root.analytics?.unavailableReason(analytic: "time_domain_hrv", name: "PRV"),
      recoveryUnavailableReason: root.analytics?.unavailableReason(analytic: "recovery", name: "Recovery"),
      hash: hash
    )
  }

  private static func prv(from analytics: HostSnapshot.Analytics) throws -> MavPrv? {
    guard let label = analytics.variabilityLabel else { return nil }
    guard let intervalSource = analytics.intervalSource,
          let meanIntervalMicros = analytics.meanIntervalMicros,
          let rmssdMicros = analytics.rmssdMicros,
          let sdnnMicros = analytics.sdnnMicros,
          let nn50Count = analytics.nn50Count,
          let pnn50MilliPercent = analytics.pnn50MilliPercent,
          let intervalCount = analytics.intervalCount,
          let excludedIntervalCount = analytics.excludedIntervalCount,
          let algorithm = analytics.algorithm,
          let algorithmVersion = analytics.algorithmVersion,
          let provenanceId = analytics.provenanceId
    else { throw MavPresentationError.incompleteAnalytics }
    return MavPrv(
      label: label,
      intervalSource: intervalSource,
      meanIntervalMicros: meanIntervalMicros,
      rmssdMicros: rmssdMicros,
      sdnnMicros: sdnnMicros,
      nn50Count: nn50Count,
      pnn50MilliPercent: pnn50MilliPercent,
      intervalCount: intervalCount,
      excludedIntervalCount: excludedIntervalCount,
      algorithm: algorithm,
      algorithmVersion: algorithmVersion,
      provenanceId: provenanceId
    )
  }
}

enum MavPresentationError: LocalizedError {
  case unsupportedSchema
  case incompleteAnalytics
  var errorDescription: String? {
    switch self {
    case .unsupportedSchema: "This version of Mav cannot read this core snapshot."
    case .incompleteAnalytics: "The core analytics snapshot is missing required fields."
    }
  }
}

private struct HostSnapshot: Decodable {
  let schema: String
  let coreVersion: String
  let storageSchema: Int
  let revision: UInt64
  let asOfUnixMs: Int64
  let connection: Connection
  let session: Session?
  let analytics: Analytics?

  enum CodingKeys: String, CodingKey {
    case schema
    case coreVersion = "core_version"
    case storageSchema = "storage_schema"
    case revision
    case asOfUnixMs = "as_of_unix_ms"
    case connection
    case session
    case analytics
  }

  struct Connection: Decodable {
    let state: String
    let displayName: String?
    let batteryPercent: Int?
    let charging: Bool?
    let onWrist: Bool?
    let lastSampleUnixMs: Int64?

    enum CodingKeys: String, CodingKey {
      case state
      case displayName = "display_name"
      case batteryPercent = "battery_percent"
      case charging
      case onWrist = "on_wrist"
      case lastSampleUnixMs = "last_sample_unix_ms"
    }
  }

  struct Session: Decodable {
    let currentBpm: Int?
    let meanMilliBpm: Int?
    let inRangeSamples: Int
    let excludedSamples: Int

    enum CodingKeys: String, CodingKey {
      case currentBpm = "current_bpm"
      case meanMilliBpm = "mean_milli_bpm"
      case inRangeSamples = "in_range_samples"
      case excludedSamples = "excluded_samples"
    }
  }

  struct Analytics: Decodable {
    let intervalSource: String?
    let variabilityLabel: String?
    let meanIntervalMicros: Int64?
    let rmssdMicros: Int64?
    let sdnnMicros: Int64?
    let nn50Count: Int?
    let pnn50MilliPercent: Int64?
    let intervalCount: Int?
    let excludedIntervalCount: Int?
    let algorithm: String?
    let algorithmVersion: String?
    let provenanceId: Int64?
    let availability: [Availability]?

    enum CodingKeys: String, CodingKey {
      case intervalSource = "interval_source"
      case variabilityLabel = "variability_label"
      case meanIntervalMicros = "mean_interval_micros"
      case rmssdMicros = "rmssd_micros"
      case sdnnMicros = "sdnn_micros"
      case nn50Count = "nn50_count"
      case pnn50MilliPercent = "pnn50_milli_percent"
      case intervalCount = "interval_count"
      case excludedIntervalCount = "excluded_interval_count"
      case algorithm
      case algorithmVersion = "algorithm_version"
      case provenanceId = "provenance_id"
      case availability
    }

    func unavailableReason(analytic: String, name: String) -> String? {
      guard let entry = availability?.first(where: { $0.analytic == analytic && !$0.available })
      else { return nil }
      return entry.reason?.message(name: name) ?? "Unavailable"
    }
  }

  struct Availability: Decodable { let analytic: String; let available: Bool; let reason: Reason? }

  struct Reason: Decodable {
    let kind: String
    let streams: [String]?

    func message(name: String) -> String {
      switch kind {
      case "algorithm_not_admitted": "\(name) model not admitted"
      case "missing_streams": "Needs \((streams ?? []).joined(separator: ", "))"
      default: "Unavailable"
      }
    }
  }
}
